src/encoder.cpp

```diff
@@ -0,0 +1,47 @@
+#include "encoder.h"
+#include "span.h"
+#include "version_number.h"
+
+#include <sstream>
+
+namespace datadog {
+namespace opentracing {
+
+// Returns a TraceEncoder that uses msgpack to produce the encoded data.
+TraceEncoder msgpackEncoder() {
+  // Allocate and reuse a buffer instead of making a new buffer per call.
+  auto buffer = std::make_shared<std::stringstream>();
+  return [=](std::deque<Trace> traces) -> std::string {
+    buffer->clear();
+    buffer->str(std::string{});
+    msgpack::pack(*buffer, traces);
+    return buffer->str();
+  };
+}
+
+AgentHttpEncoder::AgentHttpEncoder() : AgentHttpEncoder(config::tracer_version) {}
+
+AgentHttpEncoder::AgentHttpEncoder(std::string version) {
+  // Set up common headers and default encoder
+  common_headers_ = {{"Content-Type", "application/msgpack"},
+                     {"Datadog-Meta-Lang", "cpp"},
+                     {"Datadog-Meta-Tracer-Version", version}};
+  trace_encoder_ = msgpackEncoder();
+}
+
+const std::string agent_api_path = "/v0.3/traces";
+
+std::string AgentHttpEncoder::path() { return agent_api_path; }
+
+std::map<std::string, std::string> AgentHttpEncoder::headers(std::deque<Trace> &traces) {
+  std::map<std::string, std::string> headers(common_headers_);
+  headers["X-Datadog-Trace-Count"] = std::to_string(traces.size());
+  return headers;
+}
+
+std::string AgentHttpEncoder::payload(std::deque<Trace> &traces) {
+  return trace_encoder_(std::move(traces));
```

Comment:
Why do we need to move? Why not just leave it owned by Writer (since this all happens synchronously), and let writer continue to explicitly clear() it and re-use the memory. 

This should take a const reference to a Trace deque, and so should trace_encoder_.

Reply:
Yup. This was from rushing to implement. Will fix.